src/main/java/frc/robot/subsystems/vision/VisionSubsystem.java

```diff
@@ -1,175 +1,306 @@
+   * @return true if the limelight can fully see one or more April Tag
    */
-  public boolean canSeeAprilTags() {
+  public boolean canSeeAprilTags(int limelightNumber) {
     // First checks if it can see an april tag, then checks if it is fully in frame
     // Different Limelights have different FOVs
-    if (currentlyUsedLimelight.equals(VisionConstants.SHOOTER_LIMELIGHT_NAME)) {
-          return LimelightHelpers.getFiducialID(currentlyUsedLimelight) != -1
-      && Math.abs(LimelightHelpers.getTX(currentlyUsedLimelight)) <= VisionConstants.LL3G_FOV_MARGIN_OF_ERROR;
+    if (getNumberOfAprilTags(limelightNumber) > 0
+        && getNumberOfAprilTags(limelightNumber) <= VisionConstants.APRIL_TAG_POSITIONS.length) {
+      if (getLimelightName(limelightNumber).equals(VisionConstants.SHOOTER_LIMELIGHT_NAME)) {
+        return Math.abs(LimelightHelpers.getTX(getLimelightName(limelightNumber)))
+            <= VisionConstants.LL3G_FOV_MARGIN_OF_ERROR;
+      } else {
+        return Math.abs(LimelightHelpers.getTX(getLimelightName(limelightNumber)))
+            <= VisionConstants.LL3_FOV_MARGIN_OF_ERROR;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Gets the JSON dump from the specified limelight and puts it into a PoseEstimate object, which
+   * is then placed into its corresponding spot in the limelightEstimates array.
+   *
+   * @param limelightNumber the number of the limelight
+   */
+  public void updateLimelightPoseEstimate(int limelightNumber) {
+    if (!canSeeAprilTags(limelightNumber)) {
+      limelightEstimates[limelightNumber] = new PoseEstimate();
     }
-    return LimelightHelpers.getFiducialID(currentlyUsedLimelight) != -1
-      && Math.abs(LimelightHelpers.getTX(currentlyUsedLimelight)) <= VisionConstants.LL3_FOV_MARGIN_OF_ERROR;
-  }
-
-  /**
-   * Returns the pose of the robot calculated by the limelight. If there
-   * are multiple limelights that can see april tags, it uses the limelight
-   * that is closest to an april tag. 
-   */
-  public Pose2d getPoseFromAprilTags() {
-    if (canSeeAprilTags()) {
-      // This is the MT2 Code, its not
-      // LimelightHelpers.SetRobotOrientation(VisionConstants.SHOOTER_LIMELIGHT_NAME, currentRobotHeadingDegrees, 0, 0, 0, 0, 0);
-      // return LimelightHelpers.getBotPoseEstimate_wpiBlue_MegaTag2(currentlyUsedLimelight).pose;
-      Pose2d botPose = LimelightHelpers.getBotPose2d(currentlyUsedLimelight);
-      // The origin of botpose is at the center of the field
-      double robotX = botPose.getX() + FieldConstants.FIELD_LENGTH_METERS / 2.0;
-      double robotY = botPose.getY() + FieldConstants.FIELD_WIDTH_METERS / 2.0;
-      Rotation2d robotRotation = botPose.getRotation();
-      lastSeenPose = new Pose2d(robotX, robotY, robotRotation);
-      return new Pose2d(robotX, robotY, robotRotation);
+
+    // double distanceToAprilTags = getLimelightAprilTagDistance(limelightNumber);
+
+    // if (isLargeDiscrepancyBetweenMegaTag1And2(limelightNumber)
```

Comment:
I like this, we should definitely keep this in mind.

Reply:
Ishan and I are talking about this. We might want to look at what other teams are using in terms of error filtering/signal processing to filter out bad measurements. I suspect there might be algorithms out there that have already been developed to help us.